File: Tree/Binary Tree/SegmentTree/SegmentTree.cpp
#include <bits/stdc++.h>
using namespace std;

class Node {
public:
    int data;
    int start_interval, end_interval;
    Node* lchild;
    Node* rchild;

    Node(int start, int end, int val = 0) {
        start_interval = start;
        end_interval = end;
        data = val;
        lchild = rchild = nullptr;
    }
};

class Segment_Tree {
private:
    Node* construct_tree(vector<int>& arr, int left, int right) {
        if (left == right)
            return new Node(left, right, arr[left]);

        Node* node = new Node(left, right);
        int mid = (left + right) / 2;
        node->lchild = construct_tree(arr, left, mid);
        node->rchild = construct_tree(arr, mid + 1, right);
        node->data = node->lchild->data + node->rchild->data;
        return node;
    }

    int query_range(Node* node, int l, int r) {
        if (!node || l > node->end_interval || r < node->start_interval)
            return 0;  // no overlap

        if (l <= node->start_interval && r >= node->end_interval)
            return node->data;  // total overlap

        // partial overlap
        return query_range(node->lchild, l, r) + query_range(node->rchild, l, r);
    }

    void update_point(Node* node, int idx, int new_val) {
        if (!node || idx < node->start_interval || idx > node->end_interval)
            return;

        if (node->start_interval == node->end_interval) {
            node->data = new_val;
            return;
        }

        update_point(node->lchild, idx, new_val);
        update_point(node->rchild, idx, new_val);

        node->data = node->lchild->data + node->rchild->data;
    }

public:
    Node* root;

    Segment_Tree(vector<int>& arr) {
        root = construct_tree(arr, 0, arr.size() - 1);
    }

    int query(int l, int r) {
        return query_range(root, l, r);
    }

    void update(int idx, int val) {
        update_point(root, idx, val);
    }
};


int main(){
	int n,q;
	cin >> n >> q;

	vector<int> arr(n);

	for(int i=0;i<n;i++)
		cin >> arr[i];

	Segment_Tree s(arr);

	for(int i = 0 ; i < q ; i++){
		int c,l,r;

		cin >> c >> l >> r;

		if(c == 1){
			s.update(l-1, r);
		}else{
			cout << s.query(l-1, r-1) << endl;
		}
	}
	return 0;
}

File: Tree/Binary Tree/SegmentTree/SegmentTree.java
import java.util.*;
import java.io.*;

class SegmentTree{
    public class Node{
        public int data;
        public int lrange, rrange;
        Node lchild, rchild;

        public Node(int lrange, int rrange, int data){
            this.lrange = lrange;
            this.rrange = rrange;
            this.data = data;
            lchild = rchild = null;
        }
    }
    public Node head;
    Node costruct_tree(int arr[], int left, int right){
        if(left == right){
            return new Node(left, right, arr[left]);
        }

        Node root = new Node(left, right, 0);
        int mid = (left + right)/2;
        root.lchild = costruct_tree(arr, left, mid);
        root.rchild = costruct_tree(arr, mid+1, right);

        root.data = Math.min(root.lchild.data, root.rchild.data);
        return root;
    }

    int query(Node root, int left, int right){
        if(root == null || root.lrange > right || root.rrange < left)
            return Integer.MAX_VALUE;
        
        if(root.lrange >= left && root.rrange <= right)
            return root.data;
        
        return Math.min(query(root.lchild, left, right), query(root.rchild, left, right));
    }

    void update(Node root, int idx, int value){
        if(root == null || root.lrange > idx || root.rrange < idx)
            return;
        if(root.lrange == idx && root.rrange == idx){
            root.data = value;
            return;
        }

        update(root.lchild, idx, value);
        update(root.rchild, idx, value);

        root.data = Math.min(root.lchild.data, root.rchild.data);
    }

    public SegmentTree(int arr[]){
        head = costruct_tree(arr, 0 , arr.length-1);
    }

    public int query(int left, int right){
        return query(head, left, right);
    }

    public void update(int idx, int value){
        update(head, idx, value);
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        StringTokenizer st = new StringTokenizer(br.readLine());

        int n = Integer.parseInt(st.nextToken());
        int q = Integer.parseInt(st.nextToken());

        int[] arr = new int[n];
        st = new StringTokenizer(br.readLine());
        for (int i = 0; i < n; i++)
            arr[i] = Integer.parseInt(st.nextToken());

        SegmentTree seg = new SegmentTree(arr);

        for (int i = 0; i < q; i++) {
            st = new StringTokenizer(br.readLine());
            int a = Integer.parseInt(st.nextToken());
            int b = Integer.parseInt(st.nextToken());
            int c = Integer.parseInt(st.nextToken());
            if(a == 1){
                seg.update(b-1, c);
            }else sb.append(seg.query(b - 1, c - 1)).append('\n');
        }

        System.out.print(sb);
    }
}